config/crd/snapshot.storage.k8s.io_volumesnapshotclasses.yaml

```diff
@@ -0,0 +1,75 @@
+
+---
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  creationTimestamp: null
+  name: volumesnapshotclasses.snapshot.storage.k8s.io
+spec:
+  group: snapshot.storage.k8s.io
+  names:
+    kind: VolumeSnapshotClass
+    listKind: VolumeSnapshotClassList
+    plural: volumesnapshotclasses
+    singular: volumesnapshotclass
+  scope: Cluster
```

Comment:
I would strongly recommend setting [`preserveUnknownFields: false`](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#pruning-versus-preserving-unknown-fields) and making sure these CRDs are [structural](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#specifying-a-structural-schema) (creating it and checking status doesn't show a non-structural condition). That will ensure they are well-positioned to work with the v1 CRD API

Reply:
Also, since this is an official k8s.io-suffixed API, add an `api-approved.kubernetes.io` annotation containing a URL to the PR/Issue where the API is approved (see https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190612-crd-group-protection.md#summary for details). That can be this PR if this is the canonical home for this API.

Reply:
there are pending issues like [this](https://github.com/kubernetes-sigs/controller-tools/issues/279) to be resolved if kubebuilder is used. As the schemas in this PR are generated using controller-gen, I manually removed metadata field from validation to make crd valid for now.

Reply:
Is there an automatic way to validate that this schema is structural with v1beta? Or should we test it out as v1 to verify that?

Reply:
Correct me if I am wrong @msau42, you were asking to check whether the schema is valid with v1beta?
If that's your question, I have conducted manual tests on 3 different k8s versions, 1.14/1.15/1.16, all seems to be fine. v1 CRD API is stable in k8s 1.16. I have not tested using v1 though.

Reply:
I mean to test that this v1beta1 schema is compatible with v1. Can we temporarily convert this to v1 and try it on a 1.16 cluster?

Reply:
I've double verified the schemas could be seamlessly converted to v1 in a 1.16 cluster with following output.
k get crd volumesnapshotclasses.snapshot.storage.k8s.io -o yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  annotations:
    api-approved.kubernetes.io: https://github.com/kubernetes-csi/external-snapshotter/pull/139
    controller-gen.kubebuilder.io/version: (devel)
  creationTimestamp: "2019-10-24T20:37:44Z"
  generation: 1
  name: volumesnapshotclasses.snapshot.storage.k8s.io
  resourceVersion: "1238"
  selfLink: /apis/apiextensions.k8s.io/v1/customresourcedefinitions/volumesnapshotclasses.snapshot.storage.k8s.io
  uid: 524ff8f2-b783-46d4-953f-60a6fd3c32af
spec:
  conversion:
    strategy: None
  group: snapshot.storage.k8s.io
  names:
    kind: VolumeSnapshotClass
    listKind: VolumeSnapshotClassList
    plural: volumesnapshotclasses
    singular: volumesnapshotclass
  scope: Cluster
  versions:
  - name: v1beta1
    schema:
      openAPIV3Schema:
        description: VolumeSnapshotClass specifies parameters that a underlying storage
          system uses when creating a volume snapshot. A specific VolumeSnapshotClass
          is used by specifying its name in a VolumeSnapshot object. VolumeSnapshotClasses

Reply:
status:
status:
  acceptedNames:
    kind: VolumeSnapshotContent
    listKind: VolumeSnapshotContentList
    plural: volumesnapshotcontents
    singular: volumesnapshotcontent
  conditions:
  - lastTransitionTime: "2019-10-24T20:50:02Z"
    message: no conflicts found
    reason: NoConflicts
    status: "True"
    type: NamesAccepted
  - lastTransitionTime: "2019-10-24T20:50:02Z"
    message: the initial names have been accepted
    reason: InitialNamesAccepted
    status: "True"
    type: Established
  - lastTransitionTime: "2019-10-24T20:50:02Z"
    message: approved in https://github.com/kubernetes-csi/external-snapshotter/pull/139
    reason: ApprovedAnnotation
    status: "True"
    type: KubernetesAPIApprovalPolicyConformant
  storedVersions:
  - v1beta1